### ISSUE TITLE
Incorrect handling of first quantile knot in `isqf_domain_map`  of ISQF

### ISSUE THREAD

**Issue by JingangQu** [Nixtla/neuralforecast#1350]
## Description
The current implementation of `isqf_domain_map` function applies `F.softplus()` to all quantile knots, including the first one. This is incorrect because the first quantile knot represents the starting position of the distribution and should be allowed to take negative values.

## Current Implementation (Incorrect)
```python
# Line 1739 in neuralforecast/losses/pytorch.py
quantile_knots = F.softplus(input[..., start_index : start_index + num_qk]) + tol
qk_y = torch.cumsum(quantile_knots, dim=-1)
```

## Expected Behavior
The first quantile knot should remain unconstrained (can be negative), while subsequent knots should have positive increments to ensure monotonicity. This matches the reference implementation in `isqf.py`.

## Reference Implementation (Correct)
```python
quantile_knots = torch.cat(
    [
        input[..., start_index : start_index + 1],  # First knot unconstrained
        F.softplus(quantile_knots[..., start_index + 1 : start_index + num_qk]) + tol,  # Subsequent increments positive
    ],
    dim=-1,
)
qk_y = torch.cumsum(quantile_knots, dim=-1)
```

### Versions / Dependencies

Latest version of neuralforecast

### Reproduction script

Wrong code:

```python
# Line 1739 in neuralforecast/losses/pytorch.py
quantile_knots = F.softplus(input[..., start_index : start_index + num_qk]) + tol
qk_y = torch.cumsum(quantile_knots, dim=-1)
```

Correct code:

```python
quantile_knots = torch.cat(
    [
        input[..., start_index : start_index + 1],  # First knot unconstrained
        F.softplus(quantile_knots[..., start_index + 1 : start_index + num_qk]) + tol,  # Subsequent increments positive
    ],
    dim=-1,
)
qk_y = torch.cumsum(quantile_knots, dim=-1)
```

### Issue Severity

None

**Comment by elephaint** [comment]
Thanks - this is likely the issue I was encountering with ISQF, where it often failed on datasets, but I didn't find the cause yet.

I do think that in your "correct code", the second argument to torch.cat should be using the input, not quantile_knots itself, no?

**Comment by JingangQu** [comment]
Hi,

Thanks for your reply. You're absolutely right—the code should be:

```python
quantile_knots = torch.cat(
    [
        input[..., start_index : start_index + 1],  # First knot is unconstrained
        F.softplus(input[..., start_index + 1 : start_index + num_qk]) + tol,  # Ensure positive increments
    ],
    dim=-1,
)
qk_y = torch.cumsum(quantile_knots, dim=-1)
```

That said, I’ve noticed that ISQF still frequently fails during training when using the analytical CRPS loss, even after fixing the issue above.

**Comment by elephaint** [comment]
Yes I'm noticing the same in my experiments even with the fix. The fix often makes results worse too (but I agree we should stick to reference impl)

The only way I can get it to work in general is applying a Softplus on the log_prob directly, which at least ensures it always completes (i.e. enforcing positivity of the crps) and produces valid results. I'm in doubt to apply that fix, because it does make ISQF more applicable but it also strays away from reference impl. What do you think?

I'm also thinking that technically, the reference impl uses `abs` in a lot of places where we are using Softplus. The latter is obv better behaved for gradients but it's also different for negative values.

**Comment by JingangQu** [comment]
I agree that using `Softplus` is a better choice than `abs`.

> The only way I can get it to work in general is applying a Softplus on the log\_prob directly.

I noticed that `log_prob` is defined as:

```python
def log_prob(self, z: torch.Tensor) -> torch.Tensor:
    return -self.crps(z)
```

and the loss is computed as:

```python
loss_values = -distr.log_prob(y)
```

Applying `Softplus` to `log_prob` means the final loss becomes `-softplus(-self.crps(z))`. Essentially, this still encourages minimizing `crps(z)`, but the `Softplus` may help stabilize training.

In addition, I observed that when ISQF fails, the distances between `quantile_knots` often shrink to the predefined `tol`, leading to numerical issues and potentially negative CRPS values. Using `-softplus(-self.crps(z))` might mitigate this issue compared to minimizing `crps(z)` directly.

However, this is more of a workaround. Ideally, we should address the root cause — preventing `quantile_knots` from collapsing, which leads to such pathological CRPS values in the first place.

**Comment by elephaint** [comment]
Thanks, agree with all those points. The fix is harmless but doesn't address the real issue.
